man/arch/chmod.1
.TH CHMOD 1V "9 April 1986"
.\" @(#)chmod.1 1.1 92/07/30 SMI; from UCB 4.3 BSD and S5
.SH NAME
chmod \- change mode
.SH SYNOPSIS
.B chmod
[
.B \-fR
] 
.I mode 
.I filename \fB.\|.\|.\fP
.SH SYSTEM V SYNOPSIS
.B chmod
.I mode 
.I filename \fB.\|.\|.\fP
.SH DESCRIPTION
.IX  chmod  ""  "\fLchmod\fP \(em change mode"
.IX  files   "chmod command"  files   "\fLchmod\fP \(em change mode"
.IX  change  "mode of file"
.IX  change  "permissions of file"
.IX  modes  "chmod command"  modes  "\fLchmod\fP \(em change mode"
.IX  permissions  "chmod command"  permissions  "\fLchmod\fP \(em change mode"
Change the permissions, or mode, of a file or files.
Only the owner of a file (or the superuser) may change its mode.
.LP
The mode of each named file is changed according to \fImode\fP,
which may be absolute or symbolic.  
.SS Absolute Modes
An absolute \fImode\fP is an octal number constructed from the 
.SM OR 
of the following modes:
.nf
.ta +.5i
\0\fB4\fR00	read by owner
\0\fB2\fR00	write by owner
\0\fB1\fR00	execute (search in directory) by owner
.sp .25v
\00\fB4\fR0	read by group
\00\fB2\fR0	write by group
\00\fB1\fR0	execute (search) by group
.sp .25v
\000\fB4\fR	read by others
\000\fB2\fR	write by others
\000\fB1\fR	execute (search) by others
.sp .25v
\fB4\fR000	set user ID on execution
\fB2\fR000	set group ID on execution
\fB1\fR000	sticky bit, (see \fIchmod\fR\|(2) for more information)
.fi
.SS Symbolic Modes
.LP
A symbolic \fImode\fP has the form:
.IP
[ \fIwho\fP ] \fIop permission\fP [ \fIop permission\fP ]\fB.\|.\|.\fP
.LP
\fIwho\fP is a combination of: 
.ta +.5i
.nf
\fBu\fP	user's permissions
\fBg\fP	group permissions
\fBo\fP	others
\fBa\fP	all, or \fBugo\fP
.fi
.IP
If \fIwho\fR is omitted, the default is \fBa\fP,
but the setting of the file creation mask (see \fIumask\fR in
.IR sh (1)
or
.IR csh (1) 
for more information) is taken into account.  When 
.I who
is omitted,
.I chmod
will not override the restrictions of your user mask.
.LP
\fIop\fP is one of:
.ta +.5i
.nf
\fB+\fP	to add the \fIpermission\fP 
\fB\-\fP	to remove the \fIpermission\fP 
\fB=\fP	to assign the permission explicitly (all other bits for that category, owner, group, or others, will be reset).
.fi
.LP
\fIpermission\fP is any combination of:
.ta +.5i
.nf
\fBr\fP	read
\fBw\fP	write
\fBx\fP	execute
\fBX\fP	give execute permission if the file is a directory or if there is execute permission for one of the other user classes
\fBs\fP	set owner- or group-ID.  This is only useful with \fBu\fP or \fBg\fP.
\fBt\fP	set the sticky bit to save program text between processes.
.fi
.IP
The letters \fBu\fP, \fBg\fP, or \fBo\fP indicate that \fIpermission\fP
is to be taken from the current mode for the user-class.
.IP
Omitting \fIpermission\fP
is only useful with \fB=\fP, to take away all permissions.
.LP
Multiple symbolic modes, separated by commas, may be given.
Operations are performed in the order specified.  
.SH SYSTEM V DESCRIPTION
If
.I who
is omitted in a symbolic mode, it does not take the file creation mask into
account, but acts as if
.I who
were
.BR a .
The
.B \-f
and
.B \-R
flags do not apply.
.SH OPTIONS
.TP
.B \-f
Force.
.I chmod
will not complain if it fails to change the mode of a file.
.TP
.B \-R
recursively descend through directory arguments,
setting the mode for each file as described above.
.SH EXAMPLES
The first example denies write permission to others,
the second makes a file executable by all if it is executable by anyone:
.IP
\fLchmod o\-w file
.br
chmod +X file\fR
.SH "SEE ALSO"
ls(1), sh(1), csh(1), chmod(2), chown(8)


usr.bin/chmod.c
/*
 * Copyright (c) 1980 Regents of the University of California.
 * All rights reserved.  The Berkeley software License Agreement
 * specifies the terms and conditions for redistribution.
 */

#ifndef lint
static	char sccsid[] = "@(#)chmod.c 1.1 92/07/30 SMI"; /* from UCB 5.5 5/22/86 */
#endif

/*
 * chmod options mode files
 * where
 *	mode is [ugoa][+-=][rwxXstugo] or an octal number
 *	options are -Rf
 */
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/dir.h>

#define MAXRET	254	/* ensure error status modulo 256 is non-zero */

char	*modestring, *ms;
#ifndef S5EMUL
int	um;
#endif
int	status;
int	fflag;
int	rflag;

main(argc, argv)
	char *argv[];
{
	register char *p, *flags;
	register int i;
	struct stat st;

	if (argc < 3) {
		fprintf(stderr,
		    "Usage: chmod [-Rf] [ugoa][+-=][rwxXstugo] file ...\n");
		exit(-1);
	}
	argv++, --argc;
	while (argc > 0 && argv[0][0] == '-') {
		for (p = &argv[0][1]; *p; p++) switch (*p) {

		case 'R':
			rflag++;
			break;

		case 'f':
			fflag++;
			break;

		default:
			goto done;
		}
		argc--, argv++;
	}
done:
	modestring = argv[0];
#ifndef S5EMUL
	um = umask(0);
#endif
	(void) newmode(0);
	for (i = 1; i < argc; i++) {
		p = argv[i];
		/* do stat for directory arguments */
		if (lstat(p, &st) < 0) {
			status += Perror(p);
			continue;
		}
		if (rflag && (st.st_mode&S_IFMT) == S_IFDIR) {
			status += chmodr(p, st.st_mode);
			continue;
		}
		if ((st.st_mode&S_IFMT) == S_IFLNK && stat(p, &st) < 0) {
			status += Perror(p);
			continue;
		}
		if (chmod(p, newmode(st.st_mode)) < 0) {
			status += Perror(p);
			continue;
		}
	}
	if (status > MAXRET)
		status = MAXRET;
	exit(status);
	/* NOTREACHED */
}

chmodr(dir, mode)
	char *dir;
{
	register DIR *dirp;
	register struct direct *dp;
	register struct stat st;
	char savedir[1024];
	int ecode;

	ecode = 0;

	if (getwd(savedir) == 0)
		fatal(255, "%s", savedir);
	/*
	 * Change what we are given before doing it's contents
	 */
	if (chmod(dir, newmode(mode)) < 0)
		ecode += Perror(dir);

	if (chdir(dir) < 0) {
		Perror(dir);
		return (1);
	}
	if ((dirp = opendir(".")) == NULL) {
		Perror(dir);
		return (1);
	}
	dp = readdir(dirp);
	dp = readdir(dirp); /* read "." and ".." */
	for (dp = readdir(dirp); dp != NULL; dp = readdir(dirp)) {
		if (lstat(dp->d_name, &st) < 0) {
			ecode += Perror(dp->d_name);
			continue;
		}
		if ((st.st_mode&S_IFMT) == S_IFDIR) {
			ecode += chmodr(dp->d_name, st.st_mode);
			continue;
		}
		if ((st.st_mode&S_IFMT) == S_IFLNK) {
			continue;
		}
		if (chmod(dp->d_name, newmode(st.st_mode)) < 0) {
			ecode += Perror(dp->d_name);
			continue;
		}
	}
	closedir(dirp);
	if (chdir(savedir) < 0)
		fatal(255, "can't change back to %s", savedir);
	return (ecode);
}

error(fmt, a)
	char *fmt, *a;
{

	if (!fflag) {
		fprintf(stderr, "chmod: ");
		fprintf(stderr, fmt, a);
		putc('\n', stderr);
	}
	return (!fflag);
}

fatal(status, fmt, a)
	int status;
	char *fmt, *a;
{

	fflag = 0;
	(void) error(fmt, a);
	exit(status);
}

Perror(s)
	char *s;
{

	if (!fflag) {
		fprintf(stderr, "chmod: ");
		perror(s);
	}
	return (!fflag);
}

#define	USER	05700	/* user's bits */
#define	GROUP	02070	/* group's bits */
#define	OTHER	00007	/* other's bits */
#define	ALL	01777	/* all (note absence of setuid, etc) */

#define	READ	00444	/* read permit */
#define	WRITE	00222	/* write permit */
#define	EXEC	00111	/* exec permit */
#define	SETID	06000	/* set[ug]id */
#define	STICKY	01000	/* sticky bit */

newmode(nm)
	unsigned nm;
{
	register o, m, b;
	int savem;

	ms = modestring;
	savem = nm;
	m = abs();
	if (*ms == '\0') {
		/* Must use symbolic mode to change set-gid bit of directory */
		if ((nm & S_IFMT) == S_IFDIR) {
			m &= ~(SETID & GROUP);
			m |= nm & (SETID & GROUP);
		}
		return (m);
	}
	if (ms != modestring)
		fatal(255, "invalid mode");
	do {
		m = who();
		while (o = what()) {
			b = where(nm);
			switch (o) {
			case '+':
				nm |= b & m;
				break;
			case '-':
				nm &= ~(b & m);
				break;
			case '=':
				/* Use +- to change set-gid bit of directory */
				if ((nm & S_IFMT) == S_IFDIR) {
					nm &= ~(m & ~(SETID & GROUP));
					nm |= (b & ~(SETID & GROUP)) & m;
				} else {
					nm &= ~m;
					nm |= b & m;
				}
				break;
			}
		}
	} while (*ms++ == ',');
	if (*--ms)
		fatal(255, "invalid mode");
	return (nm);
}

abs()
{
	register c, i;

	i = 0;
	while ((c = *ms++) >= '0' && c <= '7')
		i = (i << 3) + (c - '0');
	ms--;
	return (i);
}

who()
{
	register m;

	m = 0;
	for (;;) switch (*ms++) {
	case 'u':
		m |= USER;
		continue;
	case 'g':
		m |= GROUP;
		continue;
	case 'o':
		m |= OTHER;
		continue;
	case 'a':
		m |= ALL;
		continue;
	default:
		ms--;
		if (m == 0)
#ifdef S5EMUL
			m = ALL;
#else
			m = ALL & ~um;
#endif
		return (m);
	}
}

what()
{

	switch (*ms) {
	case '+':
	case '-':
	case '=':
		return (*ms++);
	}
	return (0);
}

where(om)
	register om;
{
	register m;

 	m = 0;
	switch (*ms) {
	case 'u':
		m = (om & USER) >> 6;
		goto dup;
	case 'g':
		m = (om & GROUP) >> 3;
		goto dup;
	case 'o':
		m = (om & OTHER);
	dup:
		m &= (READ|WRITE|EXEC);
		m |= (m << 3) | (m << 6);
		++ms;
		return (m);
	}
	for (;;) switch (*ms++) {
	case 'r':
		m |= READ;
		continue;
	case 'w':
		m |= WRITE;
		continue;
	case 'x':
		m |= EXEC;
		continue;
	case 'X':
		if (((om & S_IFMT) == S_IFDIR) || (om & EXEC))
			m |= EXEC;
		continue;
	case 's':
		m |= SETID;
		continue;
	case 't':
		m |= STICKY;
		continue;
	default:
		ms--;
		return (m);
	}
}
